system/include/emscripten/em_asm.h

```diff
@@ -74,52 +74,44 @@
 // C++ needs to support vararg template parameter packs, e.g. like in
 // tests/core/test_em_asm_parameter_pack.cpp. Because of that, a macro-only
 // approach doesn't work (a macro applied to a parameter pack would expand
-// incorrectly). So we can use a template function instead to build a
-// std::string, and convert that to a C string.
-// String builder class is so the _sig functions can be mutually recursive.
-class __em_asm_sig_builder {
-private:
-  static char sig_char(float) {
-    return 'd';
-  }
-  static char sig_char(double) {
-    return 'd';
-  }
-  static char sig_char(int) {
-    return 'i';
-  }
-  static char sig_char(unsigned) {
-    return 'i';
-  }
-  static char sig_char(long) {
-    return 'i';
-  }
-  static char sig_char(unsigned long) {
-    return 'i';
-  }
-  template <typename T>
-  static char sig_char(T *arg) {
-    return 'i';
-  }
-
-  template <typename ...Args>
-  struct inner {
-    char buffer[sizeof...(Args)+1];
-  };
-public:
-  template <typename ...Args>
-  static const inner<Args...> __em_asm_sig(Args ...args) {
-    inner<Args...> temp;
-    char buf[sizeof...(Args)+1] = { sig_char(args)..., 0 };
-    for (int i = 0; i < sizeof...(Args)+1; ++i) {
-        temp.buffer[i] = buf[i];
-    }
-    return temp;
-  }
+// incorrectly). So we can use a template class instead to build a temporary
+// buffer of characters.
+
+// As emscripten is require to build successfully with -std=c++03, we cannot
+// use std::tuple or std::integral_constant. Using C++11 features is only a
+// warning in modern Clang, which are ignored in system headers.
+template<typename> struct __em_asm_sig {};
+template<> struct __em_asm_sig<float> { static const char value = 'd'; };
+template<> struct __em_asm_sig<double> { static const char value = 'd'; };
+template<> struct __em_asm_sig<int> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned> { static const char value = 'i'; };
+template<> struct __em_asm_sig<long> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned long> { static const char value = 'i'; };
+template<typename T> struct __em_asm_sig<T*> { static const char value = 'i'; };
+
+// Instead of std::tuple
+template<typename... Args>
+struct __em_asm_type_tuple {};
+
+// Instead of std::make_tuple
+template<typename... Args>
+__em_asm_type_tuple<Args...> __em_asm_make_type_tuple(Args... args) {
+    return {};
+}
+
+template<typename>
+struct __em_asm_sig_builder {};
+
+template<typename... Args>
+struct __em_asm_sig_builder<__em_asm_type_tuple<Args...> > {
+  char buffer[sizeof...(Args)+1] = { __em_asm_sig<Args>::value..., 0 };
```

Comment:
This should have the added benefit of reducing most of the runtime overhead currently associated with the current implementation.

Can/should this be changed to a static char* so we can invoke it as `__em_asm_sig_builder<__typeof__(__em_asm_make_type_tuple(__VA_ARGS__))>::buffer`, or does that not work?